l2geth/rollup/sync_service.go

```diff
@@ -704,10 +636,176 @@ func (s *SyncService) ApplyTransaction(tx *types.Transaction) error {
 		txRaw,
 	)
 	tx.SetTransactionMeta(newMeta)
-
 	return s.applyTransaction(tx)
 }
 
+// syncTransactionsToTip will sync all of the transactions to the tip.
+// The syncType determines the source of the transactions.
+func (s *SyncService) syncTransactionsToTip(syncType SyncType) error {
+	s.loopLock.Lock()
+	defer s.loopLock.Unlock()
+
+	for {
+		latest, err := s.client.GetLatestTransaction(syncType)
+		if err != nil {
+			return fmt.Errorf("Cannot get latest transaction: %w", err)
+		}
+		if latest == nil {
+			log.Info("No transactions to sync")
+			return nil
+		}
+		latestIndex := latest.GetMeta().Index
+		if latestIndex == nil {
+			return errors.New("Latest index is nil")
+		}
+		nextIndex := s.GetNextIndex()
+		log.Info("Syncing transactions to tip", "start", *latestIndex, "end", nextIndex)
+
+		for i := nextIndex; i <= *latestIndex; i++ {
+			tx, err := s.client.GetTransaction(i, syncType)
+			if err != nil {
+				log.Error("Cannot get latest transaction", "msg", err)
+				time.Sleep(time.Second * 2)
+				continue
+			}
+			if tx == nil {
+				return fmt.Errorf("Transaction %d is nil", i)
+			}
+			err = s.applyTransaction(tx)
+			if err != nil {
+				return fmt.Errorf("Cannot apply transaction: %w", err)
+			}
+		}
+
+		post, err := s.client.GetLatestTransaction(syncType)
+		if err != nil {
+			return fmt.Errorf("Cannot get latest transaction: %w", err)
+		}
+		postLatestIndex := post.GetMeta().Index
+		if postLatestIndex == nil {
+			return errors.New("Latest index is nil")
+		}
+		if *postLatestIndex == *latestIndex {
+			return nil
+		}
+	}
+}
+
+// syncTransactionBatchesToTip will sync all of the transaction batches to the
```

Comment:
Recommendation: Write the algorithm being implemented in the doc comment. 

Example:
1. It gets the latest transaction batch index from the DTL
2. It gets the current batch index from the database
3. It gets any batches between the two, and applies all the transactions inside that batch sequentially

Reply:
I'd like to update the spec with this algorithm - the name of the function is `syncTransactionBatchesToTip` and with the outer `while true` it is able to catch new batches that are submitted while it has been processing. The logic in the spec has assumptions that syncing to the tip actually does syncing to the tip, so in practice this handles the race condition

Reply:
https://github.com/ethereum-optimism/specs/pull/31